daemon/src/tokio_ext.rs

```diff
@@ -39,12 +40,46 @@ where
 
     fn spawn_with_handle(self) -> RemoteHandle<()>
     where
-        Self: Future<Output = ()> + Send + 'static,
+        Self: Future<Output = ()> + Send + Any + 'static,
     {
+        debug_assert!(
+            TypeId::of::<RemoteHandle<()>>() != self.type_id(),
+            "RemoteHandle<()> is a handle to already spawned task",
+        );
+
         let (future, handle) = self.remote_handle();
         // we want to disallow calls to tokio::spawn outside FutureExt
         #[allow(clippy::disallowed_method)]
         tokio::spawn(future);
         handle
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::panic;
+
+    use tokio::time::sleep;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn spawning_a_regular_future_does_not_panic() {
+        let result = panic::catch_unwind(|| {
+            let _handle = sleep(Duration::from_secs(2)).spawn_with_handle();
+        });
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn panics_when_called_spawn_with_handle_on_remote_handle() {
+        let result = panic::catch_unwind(|| {
+            let handle = sleep(Duration::from_secs(2)).spawn_with_handle();
+            let _handle_to_a_handle = handle.spawn_with_handle();
+        });
+        assert!(
+            result.is_err(),
+            "Spawning a remote handle into a separate task should panic"
+        );
+    }
```

Comment:
This test will fail if we build the tests with `--release`. I think that is fine for now but perhaps we should just replace the `debug_assert` with an actual assert?

Reply:
thanks for pointing it out! I've changed the test to only panic in debug:

```rust
+        if cfg!(debug_assertions) {
+            assert!(
+                result.is_err(),
+                "Spawning a remote handle into a separate task should panic_in_debug_mode"
+            );
+        } else {
+            assert!(result.is_ok(), "Do not panic in release mode");
+        }
```